### ISSUE TITLE
Issue with tests (libm not found by linker) when building recipe locally

### ISSUE THREAD

**Issue by maddenp** [conda-forge/eccodes-feedstock#50]
Issue:

I'm not sure that this is a bug; it could be my own ignorance -- especially about CMake, which I know almost nothing about.

I'm trying build eccodes from the feedstock locally and encounter the following error:

```
/home/pmadden/miniconda/conda-bld/eccodes_1537833700049/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: CMakeFiles/optimize_scaling.dir/optimize_scaling.c.o: undefined reference to symbol 'sqrt@@GLIBC_2.2.5'
/home/pmadden/miniconda/conda-bld/eccodes_1537833700049/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/pmadden/miniconda/conda-bld/eccodes_1537833700049/_build_env/bin/../x86_64-conda_cos6-linux-gnu/sysroot/lib/libm.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
tests/CMakeFiles/optimize_scaling.dir/build.make:88: recipe for target 'tests/optimize_scaling' failed
make[2]: *** [tests/optimize_scaling] Error 1
CMakeFiles/Makefile2:3109: recipe for target 'tests/CMakeFiles/optimize_scaling.dir/all' failed
make[1]: *** [tests/CMakeFiles/optimize_scaling.dir/all] Error 2
```

Now, if I point the feedstock's `meta.yaml` to a local copy of the eccodes 2.8.2 source where I've commented out the lines `add_subdirectory( tests )` and `add_subdirectory( examples )` in the top-level `CMakeLists.txt` file, then `conda build -c conda-forge recipe` succeeds.  (I have so far been unable to disable the tests only by making changes to the feedstock's `build.sh`.) So the problem seems to be only with the tests -- the main eccodes executables appear to link correctly.

With some help from a colleague, I was able to get the test to pass (and the whole conda build to succeed) by applying this patch to `build.sh`:

```
diff --git a/recipe/build.sh b/recipe/build.sh
index 637c517..e15ca48 100644
--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -40,13 +40,15 @@ export CFLAGS="$CFLAGS -fPIC -I$PREFIX/include"
 src_dir="$(pwd)"
 mkdir ../build
 cd ../build
+ln -s $BUILD_PREFIX/$HOST/sysroot/lib/libm.so.6 $BUILD_PREFIX/$HOST/sysroot/lib/libm.so
 cmake $src_dir \
          -DCMAKE_INSTALL_PREFIX=$PREFIX \
          -DENABLE_JPG=1 \
          -DENABLE_NETCDF=1 \
          -DENABLE_PNG=1 \
          -DENABLE_PYTHON=0 \
-         -DENABLE_FORTRAN=1
+         -DENABLE_FORTRAN=1 \
+         -DCMATH_PATH="$BUILD_PREFIX/$HOST/sysroot"
 
 make -j $CPU_COUNT >> $BUILD_OUTPUT 2>&1
 export ECCODES_TEST_VERBOSE_OUTPUT=1
```

My understanding is that the added lines here enable the `find_package( CMath )` line in the eccodes top-level `CMakeLists.txt` to find `libm`, which presumably allows the test executables to link correctly. I have no idea why this only seems to be necessary for the test executables and not the main eccodes tools.

I understand why the above works, but am certain this is anything but a solution.

Could anyone suggest a correct approach? I'd be happy to fork and try to put together a PR, but am in need of guidance.

Thanks.
<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /home/pmadden/miniconda:
#
# Name                    Version                   Build  Channel
asn1crypto                0.24.0                   py37_0  
beautifulsoup4            4.6.3                    py37_0  
ca-certificates           2018.03.07                    0  
certifi                   2018.8.24                py37_1  
cffi                      1.11.5           py37he75722e_1  
chardet                   3.0.4                    py37_1  
conda                     4.5.11                   py37_0  
conda-build               3.15.1                   py37_0  
conda-env                 2.6.0                         1  
cryptography              2.3.1            py37hc365091_0  
filelock                  3.0.8                    py37_0  
glob2                     0.6                      py37_0  
idna                      2.7                      py37_0  
jinja2                    2.10                     py37_0  
libedit                   3.1.20170329         h6b74fdf_2  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 8.2.0                hdf63c60_1  
libstdcxx-ng              8.2.0                hdf63c60_1  
markupsafe                1.0              py37h14c3975_1  
ncurses                   6.1                  hf484d3e_0  
openssl                   1.0.2p               h14c3975_0  
patchelf                  0.9                  hf484d3e_2  
pip                       10.0.1                   py37_0  
pkginfo                   1.4.2                    py37_1  
psutil                    5.4.7            py37h14c3975_0  
pycosat                   0.6.3            py37h14c3975_0  
pycparser                 2.18                     py37_1  
pyopenssl                 18.0.0                   py37_0  
pysocks                   1.6.8                    py37_0  
python                    3.7.0                hc3d631a_0  
pytz                      2018.5                   py37_0  
pyyaml                    3.13             py37h14c3975_0  
readline                  7.0                  h7b6447c_5  
requests                  2.19.1                   py37_0  
ruamel_yaml               0.15.46          py37h14c3975_0  
setuptools                40.2.0                   py37_0  
six                       1.11.0                   py37_1  
sqlite                    3.24.0               h84994c4_0  
tk                        8.6.8                hbc83047_0  
tqdm                      4.26.0           py37h28b3542_0  
urllib3                   1.23                     py37_0  
wheel                     0.31.1                   py37_0  
xz                        5.2.4                h14c3975_4  
yaml                      0.1.7                had09818_2  
zlib                      1.2.11               ha838bed_2  
```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info

     active environment : None
            shell level : 0
       user config file : /home/pmadden/.condarc
 populated config files : 
          conda version : 4.5.11
    conda-build version : 3.15.1
         python version : 3.7.0.final.0
       base environment : /home/pmadden/miniconda  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/linux-64
                          https://repo.anaconda.com/pkgs/pro/noarch
          package cache : /home/pmadden/miniconda/pkgs
                          /home/pmadden/.conda/pkgs
       envs directories : /home/pmadden/miniconda/envs
                          /home/pmadden/.conda/envs
               platform : linux-64
             user-agent : conda/4.5.11 requests/2.19.1 CPython/3.7.0 Linux/4.15.0-34-generic ubuntu/18.04 glibc/2.27
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
</details>

**Comment by kmuehlbauer** [comment]
@maddenp Can you please make the whole conda build log available? I do not have any issues when building locally. How do you initiate the build process?

If you just use `conda build -c conda-forge recipe` the pinning would be lost, IIRC. Can you try with

`conda build recipe -m .ci_support/linux_.yaml`

**Comment by maddenp** [comment]
@kmuehlbauer thank you, yes, the build completes if I initiate it with `conda build -c conda-forge recipe -m .ci_support/linux_.yaml`.

I'd like to understand what's happening here. Do you think it's the version pinning on hdf5 and the other packages in `linux_.yaml` that is crucial? If so (although I understand this wouldn't be a cross-platform solution), would I likely also be able to build by doing those pins directly in `meta.yaml`? Or is it the `toolchain_*` definitions that are crucial? Apart from the conda-build docs on [Build Variants](https://conda.io/docs/user-guide/tasks/build-packages/variants.html), which I'll study, are there any docs specific to conda-forge on this topic that I should read?

I notice in the successful build a number of overlinking warnings and suspect (I'll test) that this build would not succeed with `--error-overlinking` specified -- even though, as I mentioned, it did succeed when I disabled those tests in eccodes. Is the overlinking desirable/acceptable/unavoidable?

Finally, and I'm sorry for all the questions, the ultimate goal here (I'm working with @t-brown, who you helped with an issue a few days ago) is to get eccodes built with the 7.3.0 compilers in order to get gfortran support for some newer Fortran features in local code we'll link to eccodes. We wanted to get that working locally before asking for / suggesting an update, but I'm uncertain how to do it. Any advice?

Thanks again for your time and support.
[build.log](https://github.com/conda-forge/eccodes-feedstock/files/2415350/build.log)

**Comment by maddenp** [comment]
Trying to understand this, I've made a trivial package (attached) and find that when I build it with `conda build -c conda-forge --error-overlinking recipe` it builds fine, with `FC` set to the expected Anaconda compiler, no overlinking errors, etc. However, when I build wth `conda build -c conda-forge --error-overlinking -m recipe/linux_.yaml recipe`, it appears that the Anaconda compilers are no longer installed, `FC` is set to just `gfortran` (which, on my local system, is the OS-supplied `/usr/bin/gfortran`), and the build halts on overlinking errors.

Is this build configuration only meant to work in specialized environments where the Anaconda compilers are already installed in the root environment so that e.g. `FC` refers to `gfortran_impl_linux-64`? That would make sense. In any case, I don't see the kind of overlinking warnings in the Circle CI output for the latest eccodes build that I saw on my local system, so perhaps the build environment there doesn't have an incorrect system gfortran to accidentally detect.

[foo.tar.gz](https://github.com/conda-forge/eccodes-feedstock/files/2416868/foo.tar.gz)

I don't think there's an eccodes-feedstock problem here. I'd like to understand the purpose and function of the `toolchain_*` packages, but is there somewhere more appropriate to ask or learn about that?

**Comment by kmuehlbauer** [comment]
@maddenp `conda-forge` has it's own ecosystem of compilers. Currently we are in the process of transition to use the new compilers. But that is slowly progressing. As `eccodes` doesn't have much of dependencies it might be worth to try to use the new compilers.

If I understood @ocefpaf  correctly, he will give it a try when he finds the time. He can also point you to the needed resources. I'm not as close to this compiler discussion. You might also ask at the conda-forge gitter channel.

**Comment by maddenp** [comment]
Thanks again @kmuehlbauer. I'll follow up on the gitter channel (I'd forgotten that exists). If @ocefpaf points me to any resources, I'll look into those, too, and hopefully be able to make useful contributions in the future.

**Comment by ocefpaf** [comment]
@maddenp I was going to test it locally only, sorry for the lack of clarity.

The reason is b/c we need to rebuild all the dependencies with the new compilers and the bot is doing that already. So it will happen here automatically at some point. My goal was just to verify if we are going to hit any compiler issues and trying to get ahead of the bot with a possible fix.

TL;DR I don't really recommend you using your time for that unless you really want to do it.
